src/libraries/Common/src/System/Reflection/Metadata/AssemblyNameInfo.cs

```diff
@@ -0,0 +1,222 @@
+    /// It's a more lightweight, immutable version of <seealso cref="AssemblyName"/> that does not pre-allocate <seealso cref="System.Globalization.CultureInfo"/> instances.
+    /// </remarks>
+    [DebuggerDisplay("{FullName}")]
+#if SYSTEM_PRIVATE_CORELIB
+    internal
+#else
+    public
+#endif
+    sealed class AssemblyNameInfo
+    {
+        internal readonly AssemblyNameFlags _flags;
+        private string? _fullName;
+
+#if !SYSTEM_PRIVATE_CORELIB
+        /// <summary>
+        /// Initializes a new instance of the AssemblyNameInfo class.
+        /// </summary>
+        /// <param name="name">The simple name of the assembly.</param>
+        /// <param name="version">The version of the assembly.</param>
+        /// <param name="cultureName">The name of the culture associated with the assembly.</param>
+        /// <param name="flags">The attributes of the assembly.</param>
+        /// <param name="publicKeyOrToken">The public key or its token. Set <paramref name="flags"/> to <seealso cref="AssemblyNameFlags.PublicKey"/> when it's public key.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="name"/> is null.</exception>
+        public AssemblyNameInfo(string name, Version? version = null, string? cultureName = null, AssemblyNameFlags flags = AssemblyNameFlags.None, ImmutableArray<byte> publicKeyOrToken = default)
+        {
+            Name = name ?? throw new ArgumentNullException(nameof(name));
+            Version = version;
+            CultureName = cultureName;
+            _flags = flags;
+            PublicKeyOrToken = publicKeyOrToken;
+        }
+#endif
+
+        internal AssemblyNameInfo(AssemblyNameParser.AssemblyNameParts parts)
+        {
+            Name = parts._name;
+            Version = parts._version;
+            CultureName = parts._cultureName;
+            _flags = parts._flags;
+#if SYSTEM_PRIVATE_CORELIB
+            PublicKeyOrToken = parts._publicKeyOrToken;
+#else
+            PublicKeyOrToken = parts._publicKeyOrToken is null ? default : parts._publicKeyOrToken.Length == 0
+                ? ImmutableArray<byte>.Empty
+    #if NET8_0_OR_GREATER
+                : Runtime.InteropServices.ImmutableCollectionsMarshal.AsImmutableArray(parts._publicKeyOrToken);
+    #else
+                : ImmutableArray.Create(parts._publicKeyOrToken);
+    #endif
+#endif
+        }
+
+        /// <summary>
+        /// Gets the simple name of the assembly.
+        /// </summary>
+        public string Name { get; }
+
+        /// <summary>
+        /// Gets the version of the assembly.
+        /// </summary>
+        public Version? Version { get; }
+
+        /// <summary>
+        /// Gets the name of the culture associated with the assembly.
+        /// </summary>
+        public string? CultureName { get; }
+
+        /// <summary>
+        /// Gets the attributes of the assembly.
+        /// </summary>
+        public AssemblyNameFlags Flags => ExtractAssemblyNameFlags(_flags);
```

Comment:
It looks odd that this public property does not just return the flags that were set by the public constructor.